### PR TITLE
[quantization] Quantization of Llama

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR quantizes the full `LLama` model and converts it to circle format.

<details> <summary> Log of `python tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py --model Maykeye/TinyLLama-v0 --save_circles_to_folder  "."  --max_seq_len 2048` </summary>

```

Namespace(model='Maykeye/TinyLLama-v0', device='cuda', dtype='float32', seed=42, trust_remote_code=False, hf_token=None, no_tqdm=False, no_GPTQ=False, no_PTQ=False, save_circle_to_folder='.', cache_dir='/mnt/storage/transformers_cache', nsamples_for_qcalibration=128, linear_weight_bits=4, gptq_mse=False, max_seq_len=2048, embedding_weight_bits=8, lm_head_weight_bits=4, eval_tasks=None)
=== Config ===
Model            : Maykeye/TinyLLama-v0
Device           : cuda
DType            : float32

Loading FP model …
You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565 - if you loaded a llama tokenizer from a GGUF file you can ignore this message.

Calculating original perplexities …
Token indices sequence length is longer than the specified maximum sequence length for this model (324381 > 2048). Running this sequence through the model will result in indexing errors
PPL:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌ | 158/159 [00:04<00:00, 37.60it/s]

┌── Wikitext-2 test perplexity ─────────────
│ FP32 :  7584.31
└───────────────────────────────────────────
Applying GPTQ …
Quantizing layers: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:08<00:00,  1.07s/layer]
Wrapping layers with PTQWrapper …                                                                                                                                                                                                                                                       
Calibrating PTQ obeservers…
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:31<00:00,  4.12it/s]

Calculating perplexities …
PPL:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌ | 158/159 [00:35<00:00,  4.47it/s]

┌── Wikitext-2 test perplexity ─────────────
│ int16 :  7410.80
└───────────────────────────────────────────
saving input embedding to /mnt/storage/slow_repos/TICO/embedding.q.circle
saving lm_head to /mnt/storage/slow_repos/TICO/lm_head.q.circle
saving layers
saving model layer_0 to /mnt/storage/slow_repos/TICO/decoder_layer_0.q.circle
saving model layer_1 to /mnt/storage/slow_repos/TICO/decoder_layer_1.q.circle
saving model layer_2 to /mnt/storage/slow_repos/TICO/decoder_layer_2.q.circle
saving model layer_3 to /mnt/storage/slow_repos/TICO/decoder_layer_3.q.circle
saving model layer_4 to /mnt/storage/slow_repos/TICO/decoder_layer_4.q.circle
saving model layer_5 to /mnt/storage/slow_repos/TICO/decoder_layer_5.q.circle
saving model layer_6 to /mnt/storage/slow_repos/TICO/decoder_layer_6.q.circle
saving model layer_7 to /mnt/storage/slow_repos/TICO/decoder_layer_7.q.circle
saving model.model to /mnt/storage/slow_repos/TICO/model.model.q.circle
saving the whole model to /mnt/storage/slow_repos/TICO/model.q.circle

```

</details>

Draft: #436
TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Comment by mhs4670go** [comment]
> Should i provide tests for it and/or split in smaller PRs?

No you don't have to. The exmaple scirpt can be a test by itself.

**Comment by stamalakhov** [comment]
@mhs4670go 
not sure about CI failure `check-style`.

**Comment by mhs4670go** [comment]
> not sure about CI failure check-style.

This is a problem of UFMT. It uses deprecated feature in Python 3.14. Let's try to use ufmt==2.9.1. I'll post a PR for this.

**Comment by mhs4670go** [comment]
@stamalakhov Could you rebase the PR?

**Comment by stamalakhov** [comment]
> @stamalakhov Could you rebase the PR?

@mhs4670go 
Rebased, but still no luck :cry:

File: tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py [hunk 1]
```diff
@@ -185,9 +185,9 @@ def forward(
         residual = hidden_states
         hidden_states = self.input_layernorm(hidden_states)
 
-        if attention_mask is None or attention_mask.dtype == torch.bool:
-            L = hidden_states.size(1)
-            attention_mask = self._slice_causal(L, hidden_states.device)
+        # to prevent introduction of attention_mask as a parameter let's use preset attention_mask
+        L = hidden_states.size(1)
+        attention_mask = self._slice_causal(L, hidden_states.device)
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: Right now. Without setting explicit attention mask to model, `LlamaModel` sets its own default casual mask and it has wrong dimensions. 
I'll try to set it explicitely in conversion.

[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: @mhs4670go 
This change is not necessary. I'll remove it. Thank you!

[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_decoder_layer.py]: @mhs4670go 
In case attention_mask is set explicitely like this:

```
attention_mask = torch.ones(1, q_m.config.max_position_embeddings, dtype=torch.bool)
cm = tico.convert(q_m, (calib_inputs[0], attention_mask), strict=False)
cm.save(save_path)
```
`_prepare_4d_causal_attention_mask_with_cache_position` of `transformers` will turn it into `floats`. So we need to quantize it inside `QuantLLamaDecoderLayer`
